### PR TITLE
Cmsjoin

### PR DESCRIPTION
**Pull request by ravwojdyla**

### CODE REVIEW

**Comment by ravwojdyla** [comment]
related to #9

**Comment by ravwojdyla** [comment]
this is rfc - do not merge - still need to add some comments and fix style here and there, so please comment on the idea/implementation.

**Comment by ravwojdyla** [comment]
ready for review @andrewsmartin @nevillelyh

File: scio-core/src/main/scala/com/spotify/scio/util/FunctionsWithSideInput.scala [hunk 1]
```diff
@@ -54,4 +54,24 @@ private[scio] object FunctionsWithSideInput {
       c.output(g(c.element(), sideInputContext(c)))
   }
 
+  /**
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/util/FunctionsWithSideInput.scala]: Now that we are not exposing this, maybe move this logic to the same place in PairSCollectionFunctions so it's more coherent?

File: scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala [hunk 2]
```diff
@@ -193,6 +194,108 @@ class PairSCollectionFunctions[K, V](val self: SCollection[(K, V)])
...
+  : SCollection[(K, (V, Option[W]))] = self.transform { in =>
+    val side = that.asMultiMapSideInput
+    in.withSideInputs(side).flatMap[(K, (V, Option[W]))] { (kv, s) =>
+      val (k, v) = kv
+      val m = s(side)
+      if (m.contains(k)) m(k).map(w => (k, (v, Some(w)))) else Seq((k, (v, None)))
+    }.toSCollection
+  }
+
+  /* Skew operations */
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: Don't really need this since there's only one method in this section?

File: scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala [hunk 3]
```diff
@@ -193,6 +194,108 @@ class PairSCollectionFunctions[K, V](val self: SCollection[(K, V)])
...
+                                delta: Double = 1E-10,
+                                seed: Int = 42)(implicit hasher: CMSHasher[K])
+  : SCollection[(K, (V, W))] = {
+    import com.twitter.algebird._
+
+    // Count k->#values:
+    val keyAggregator = CMS.aggregator[K](eps, delta, seed)
+
+    // Use asIterableSideInput as workaround for:
+    // http://stackoverflow.com/questions/37126729/ismsinkwriter-expects-keys-to-be-written-in-strictly-increasing-order
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: This will freak scalastyle out. Suppress?

File: scio-core/src/main/scala/com/spotify/scio/values/SCollectionWithSideInput.scala [hunk 4]
```diff
@@ -67,6 +68,28 @@ class SCollectionWithSideInput[T: ClassTag] private[values] (val internal: PColl
     new SCollectionWithSideInput[U](o, context, sides)
   }
 
+  /**
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/values/SCollectionWithSideInput.scala]: Same here, move the logic to `skewedN1Join`?

File: scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala [hunk 5]
```diff
@@ -193,6 +194,108 @@ class PairSCollectionFunctions[K, V](val self: SCollection[(K, V)])
...
+    val keyCMS = self.keys.aggregate(keyAggregator).asIterableSideInput
+
+    val (hotSelf, chillSelf) = (SideOutput[(K,V)](), SideOutput[(K,V)]())
+    val partitionedSelf = self
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: without re-implementing private `partition`, maybe you can do `self.cross(keyCMS)` and then normal withSideOutputs?

File: scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala [hunk 6]
```diff
@@ -193,6 +194,108 @@ class PairSCollectionFunctions[K, V](val self: SCollection[(K, V)])
...
+   * @param delta A bound on the probability that a query estimate does not lie within some small
+   *              interval (an interval that depends on `eps`) around the truth.
+   * @param seed A seed to initialize the random number generator used to create the pairwise
+   *             independent hash functions.
+   */
+  def skewedN1Join[W: ClassTag](that: SCollection[(K, W)],
+                                hotKeyThreshold: Long = 9000,
+                                eps: Double = 0.0001,
+                                delta: Double = 1E-10,
+                                seed: Int = 42)(implicit hasher: CMSHasher[K])
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: Seed should default to something random and hard coded in a test?

File: scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala [hunk 7]
```diff
@@ -193,6 +194,108 @@ class PairSCollectionFunctions[K, V](val self: SCollection[(K, V)])
...
+      val (k, v) = kv
+      val m = s(side)
+      if (m.contains(k)) m(k).map(w => (k, (v, Some(w)))) else Seq((k, (v, None)))
+    }.toSCollection
+  }
+
+  /* Skew operations */
+
+  /**
+   * N to 1 skewproof flavor of [[PairSCollectionFunctions.join()]].
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: Do we really need N1 in the name? Maybe hard to explain in layman's them. `skewedJoin` sounds clear enough?

File: scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala (hunk 3 above)
Comment (ravwojdyla) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: done.

File: scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala (hunk 5 above)
Comment (ravwojdyla) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: Isn't this gonna blow up left site? If I understand it correctly it will duplicate whole CMS for each record of the left side. While partition is basically a filter on CMS?

File: scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala (hunk 6 above)
Comment (ravwojdyla) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: Something `random` - you mean more random than `42`, or remove all default values and give example in the doc? After thinking about it a bit - maybe it's better to remove the default all together and give example in doc. ?

File: scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala (hunk 7 above)
Comment (ravwojdyla) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: Name changed to `skewedJoin`.

File: scio-core/src/main/scala/com/spotify/scio/values/SCollectionWithSideInput.scala (hunk 4 above)
Comment (ravwojdyla) [inline scio-core/src/main/scala/com/spotify/scio/values/SCollectionWithSideInput.scala]: Would prefer to leave it here to have access to `parDo` etc
